plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+            logInfo . render $ "Updated contract" <+> pretty instanceID <+> "to new iteration" <+> pretty newIteration
+        _ -> logInfo "No messages."
+    logInfo "processContractInbox end"
+
+-- | Check the inboxes of all contracts.
+processAllContractInboxes ::
+    forall t effs.
+        ( Member (EventLogEffect (ChainEvent t)) effs
+        , Member (ContractEffect t) effs
+        , Member (Error SCBError) effs
+        , Member Log effs
+        )
+    => Eff effs ()
+processAllContractInboxes = do
+    logInfo "processAllContractInboxes start"
+    ContractInboxState{cisInboxMessages} <- runGlobalQuery (Query.inboxMessages @t)
+    traverse_ (uncurry (processContractInbox @t)) (Map.toList cisInboxMessages)
+    logInfo "processAllContractInboxes end"
+
+-- TODO different return type to indicate that the contract
+-- does indeed exist
+lookupContract ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member (Error SCBError) effs
+    , Show t
+    , Ord t
+    )
+    => t
+    -> Eff effs t
+lookupContract t = do
+    installed <- Projections.installedContracts
+    let matchingContracts =
+            Set.filter
+                (\cp -> cp == t)
+                installed
+    case Set.lookupMin matchingContracts of
+        Just c  -> pure c
+        Nothing -> throwError (ContractNotFound (show t))
+
+-- | Create a new instance of the contract
+activateContract ::
+    forall t effs.
+    ( Member Log effs
+    , Member (EventLogEffect (ChainEvent t)) effs
+    , Member (Error SCBError) effs
+    , Member UUIDEffect effs
+    , Member (ContractEffect t) effs
+    , Show t
+    , Pretty t
+    , Ord t
+    )
+    => t
+    -> Eff effs ContractInstanceId
+activateContract contract = do
+    logInfo . render $ "Finding contract" <+> pretty contract
+    contractDef <- lookupContract @t contract
+    activeContractInstanceId <- ContractInstanceId <$> uuidNextRandom
+    logInfo . render $ "Initializing contract instance with ID" <+> pretty activeContractInstanceId
+    let currentIteration = iterationZero
+    response <- Contract.invokeContract @t $ InitContract contractDef
+    sendContractStateMessages @t contract activeContractInstanceId currentIteration response
+    logInfo . render $ "Activated contract instance:" <+> pretty activeContractInstanceId
+    pure activeContractInstanceId
+
+processOwnPubkeyRequests ::
+    forall t effs.
+    ( Member Log effs
+    , Member (EventLogEffect (ChainEvent t)) effs
+    , Member WalletEffect effs
+    )
+    => Eff effs ()
+processOwnPubkeyRequests = do
+    logInfo "processOwnPubkeyRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.ownPubKeyRequests @t)
+    let go pk (i, (it, ())) = do
+            logDebug . render $ "processOwnPubkeyRequest" <+> pretty i <+> pretty pk
+            sendContractInboxMessage @t (OwnPubkeyResponse pk) i it
+    unless (Map.null state) $ do
```

Comment:
This only saves us the `ownPubKey` lookup, right?

Reply:
Likewise in subsequent places (as in, the traverse will already do nothing if the map is empty, but the lookup does). So maybe just worth a comment.

Reply:
Done.